scripts: new log

georglauterbach opened the pull request:
# Description

This PR introduces a new `_log` function. As a successor to the closed #2486, this PR is still big, but tighter and only brings

1. the new `_log` function
2. updated log levels for most scripts
3. necessary documentation updates

## Type of change

<!-- Delete options that are not relevant. -->

- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

georglauterbach commented:
@polarathene @casperklein I hope this PR is easier to review. It does not contain changes to `check-for-changes.sh`, this will be done in the future.

polarathene commented on docs/content/contributing/coding-style.md:
```diff
@@ -115,7 +115,7 @@ A positive example, which is taken from `setup-stack.sh`, would be
 function _setup_postfix_aliases
 {
-  _notify 'task' 'Setting up Postfix Aliases'
+  _log 'task' 'Setting up Postfix Aliases'
```
```suggestion
  _log 'debug' 'Setting up Postfix aliases'
```

georglauterbach replied:
👍🏼

georglauterbach replied:
Done in f097b1d08e40259d9d558b79a9875518292a9466

polarathene commented on docs/content/contributing/coding-style.md:
```diff
@@ -142,7 +142,7 @@ function _setup_postfix_aliases
     done < <(grep -v "^\s*$\|^\s*\#" /tmp/docker-mailserver/postfix-virtual.cf || true)
   else
-    _notify 'inf' "Warning '/tmp/docker-mailserver/postfix-virtual.cf' is not provided. No mail alias/forward created."
+    _log 'inf' "Warning '/tmp/docker-mailserver/postfix-virtual.cf' is not provided. No mail alias/forward created."
```
```suggestion
    _log 'debug' "'/tmp/docker-mailserver/postfix-virtual.cf' not provided - no mail alias/forward created"
```

georglauterbach replied:
👍🏼

georglauterbach replied:
Done in Done in f097b1d08e40259d9d558b79a9875518292a9466

casperklein commented on target/scripts/start-mailserver.sh:
```diff
@@ -260,20 +262,20 @@ source /usr/local/bin/daemons-stack.sh
 date >/CONTAINER_START
 
-_notify 'tasklog' "${HOSTNAME} is up and running"
+echo -e "\n${HOSTNAME} is up and running\n"
```
```suggestion
_log 'info' "${HOSTNAME} is up and running"
```

polarathene replied:
Another carry-over missing the commit message context ([original commit with context](https://github.com/docker-mailserver/docker-mailserver/pull/2486/commits/16b740e84e7433557cb6674ad601331adea21263)).

During original review, @georglauterbach added an `always` log level for log content that should persist through all levels, even when it wasn't fitting for that level. Such as an `info` that should log when using `error` or `warn` levels.

He later decided to drop the `always` log level, and just use a plain echo for such cases instead. I guess that has a little risk of other maintainers not knowing that and accidentally correcting to `info`, unless you think this should not be logged in `error` and `warn`?

georglauterbach replied:
@polarathene is right, I want these messages to always be logged. I will add a comment to signal that these two are the only exceptions where `echo` is used instead of `_log`.

georglauterbach replied:
Done in 2b9ab7c52949678ccdfc773a65744015c8cc989d

casperklein replied:
Alright, but then it doesn't align and don't look uniformly/pretty:

```
[  TRACE  ]  Some very verbose trace message
[  DEBUG  ]  Some verbose debug message
[   INF   ]  Some info message

${HOSTNAME} is up and running

[ WARNING ]  Some warning message
[  ERROR  ]  Some error message
```

Then manually add INFO? `echo "[   INF   ]  ${HOSTNAME} is up and running"`

```
[  TRACE  ]  Some very verbose trace message
[  DEBUG  ]  Some verbose debug message
[   INF   ]  Some info message
[   INF   ]  ${HOSTNAME} is up and running
[ WARNING ]  Some warning message
[  ERROR  ]  Some error message
```

The same goes for the "welcome to DMS" message. We must decide, if we want a continuous look & feel, or not. I prefer a clean look.

georglauterbach replied:
I would normally go for the second variant, but then it may be odd, especially when you set the log level to `warn` or `error`, that in info message is logged.

Therefore, I made a decision that should be comforting for all of us: I revised my earlier idea, and we will just log the message normally with `_log` on the `info` message. After all, these messages carry informational output, and they're not particularly special (after thinking about it twice; my idea was to always log them because they are in `start-mailserver.sh`). I hope this is now okay for everyone :)

georglauterbach replied:
Done in [`1cbde1d`](https://github.com/docker-mailserver/docker-mailserver/pull/2493/commits/1cbde1d139f519716d0c439eb871fe7c37ded85d) and [`c098617`](https://github.com/docker-mailserver/docker-mailserver/pull/2493/commits/c098617f83d1fa2e1c3e0c9049267026b4b2e176)

polarathene reviewed (dismissed):
Excellent work :+1: 

Nicely organized commits, easy to review diffs/changes. Thanks! :grinning: 

Only hiccup I noticed was `coding-style.md` doc log levels. Not a major issue as the `setup-stack.sh` snippet is now outdated (_bulk of it moved to `aliases.sh` helper_). I added two commit suggestions that just apply the changes you made to those two log lines. No need to further adjust the snippet for this PR, that can always be done in a small docs fix PR if syncing it to latest content has any value.

polarathene commented on target/scripts/startup/setup-stack.sh:
```diff
@@ -271,9 +272,9 @@ function _setup_dovecot_quota
         /etc/dovecot/conf.d/90-quota.conf
 
-      if [[ ! -f /tmp/docker-mailserver/dovecot-quotas.cf ]]
+      if [[ -d /tmp/docker-mailserver ]] && [[ ! -f /tmp/docker-mailserver/dovecot-quotas.cf ]]
```
This was covered in the original PR review. Apparently if the directory doesn't exist, it fails?

https://github.com/docker-mailserver/docker-mailserver/pull/2486/commits/567595878e77b03d6ab16c3f4f4554d5083de77a#diff-141cf3167e0d3e34aa4f5242772a4e7519a0abb8a7cd2ba4493e8ea0ed607a61

> 4. Corrected a bug in `setup-stack.sh` where the parent directory would not be present and the following `: >...` would not succeed

georglauterbach replied:
When DMS starts, but has faulty mounts, an error is shown here (although a real `error` messages was logged before). The command below `: >/tmp/docker-mailserver/dovecot-quotas.cf` will not create the directories and show an error.

This is however a leftover from a simple copy. `setup-stack.sh` is still over a thousand lines long and I missed this. This addition does not belong in this PR and I will remove it.

georglauterbach replied:
Done in f097b1d08e40259d9d558b79a9875518292a9466

polarathene commented on target/scripts/helpers/lock.sh:
```diff
@@ -12,19 +12,22 @@ function _create_lock
   while [[ -e "${LOCK_FILE}" ]]
   do
-    _notify 'warn' "Lock file ${LOCK_FILE} exists. Another ${SCRIPT_NAME} execution is happening. Trying again shortly..."
+    _log 'warn' \
+      "Lock file ${LOCK_FILE} exists. Another ${SCRIPT_NAME} execution is happening -" \
+      'trying again shortly'
```
It used to allow more than 1 arg for message, I didn't see the point of it I think and requested it be changed to a fixed two. This wasn't something I realized would be an issue here, I just assumed it was like a multi-line string for some reason :sweat_smile:

georglauterbach replied:
This is an example of why I wanted to use `$*` instead of `$2`, but I guess using a one-liner is fine here. I will change it.

georglauterbach replied:
Done in 2b9ab7c52949678ccdfc773a65744015c8cc989d

georglauterbach commented on target/scripts/helpers/log.sh:
```diff
@@ -1,5 +1,103 @@
+# is missing. Both failures will return with exit code '1'.
+function _log
+{
+  if [[ -z ${1+set} ]]
```
`+` expands if the variable is set. I'm using this construct everywhere I check whether a variable is set, and according to [this SO post](https://stackoverflow.com/questions/3601515/how-to-check-if-a-variable-is-set-in-bash), this is the best way of doing it.

While `:-` would be very much fine as well (for this use case!), there is no need in changing it :)

casperklein replied:
I understand, what `:+` does 😉 I just wondered, what the benefit in returning a string to the test function is, compared to return nothing like we otherwise do. Looks like overhead to me.

PS: `$1` would also be enough, as we don't use `set -u` in any script, that does source this function.

georglauterbach replied:
Yes, adding `set -u` to `start-mailserver.sh` would be nice! You're right, this must be done carefully (and in another PR of course 😉).

georglauterbach replied:
I will resolve this because I like the `+` and we can afford 3 more bytes in RAM :D

casperklein replied:
ack

georglauterbach commented on target/scripts/helpers/log.sh:
```diff
@@ -1,5 +1,103 @@
+
+    ( 'info' )
+      [[ ${LEVEL_AS_INT} -ge 3 ]] || return 0
+      MESSAGE+="   ${LOG_BLUE}INF   "
```
1. `_notify` uses `INF` as well :)
2. This would break my much-loved symmetry

I think you have a small misconception about the message format I chose (seeing https://github.com/docker-mailserver/docker-mailserver/pull/2493#discussion_r830541017 as well): With this PR, log looks like this:

``` TXT
[  TRACE  ]  Some very verbose trace message
[  DEBUG  ]  Some verbose debug message
[   INF   ]  Some info message
[ WARNING ]  Some warning message
[  ERROR  ]  Some error message
```

I can change it to `INFO` if this is desired by a majority (or if we decide it at a later point in time), but I'd like to keep it the way it is (at least for now) :) This also requires the spaces around the strings (these are not there by accident) :D

georglauterbach commented on target/scripts/helpers/log.sh:
```diff
@@ -1,5 +1,103 @@
+  fi
+
+  local MESSAGE LEVEL_AS_INT
+  MESSAGE="${LOG_RESET}["
```
See my comment above that resolves this: https://github.com/docker-mailserver/docker-mailserver/pull/2493#discussion_r830577356

georglauterbach commented on target/scripts/helpers/log.sh:
```diff
@@ -1,5 +1,103 @@
+  case "${1}" in
+    ( 'trace' )
+      [[ ${LEVEL_AS_INT} -ge 5 ]] || return 0
+      MESSAGE+="  ${LOG_LGRAY}TRACE  "
```
https://github.com/docker-mailserver/docker-mailserver/pull/2493#discussion_r830577356

https://github.com/docker-mailserver/docker-mailserver/pull/2493#discussion_r830577405

georglauterbach commented on target/scripts/helpers/log.sh:
```diff
@@ -1,5 +1,103 @@
+      ;;
+  esac
+
+  shift 1
+  MESSAGE+="${LOG_RESET}]  ${*}"
```
I will change this with an upcoming commit 👍🏼

georglauterbach replied:
Done in 2b9ab7c52949678ccdfc773a65744015c8cc989d

georglauterbach commented on docs/content/contributing/coding-style.md:
```diff
@@ -142,7 +142,7 @@ function _setup_postfix_aliases
     done < <(grep -v "^\s*$\|^\s*\#" /tmp/docker-mailserver/postfix-virtual.cf || true)
   else
-    _notify 'inf' "Warning '/tmp/docker-mailserver/postfix-virtual.cf' is not provided. No mail alias/forward created."
+    _log 'inf' "Warning '/tmp/docker-mailserver/postfix-virtual.cf' is not provided. No mail alias/forward created."
```
Yes, I overlooked this. @polarathene has also noted this.

I usually changed `inf` to `debug` or `trace` and will do so with the next upcoming commit as well :)

georglauterbach replied:
Done in 2b9ab7c52949678ccdfc773a65744015c8cc989d

georglauterbach commented on target/scripts/helpers/error.sh:
```diff
@@ -2,7 +2,8 @@
 function _errex
 {
-  echo -e "Error :: ${*}\nAborting." >&2
+  _log 'error' "${1:?}"
```
`:?` has nothing to with where the message it logged to. It makes sure the first argument is properly supplied. Otherwise `_log` would log an error that no message is provided, which would be misleading.

casperklein replied:
> Otherwise _log would log an error that no message is provided, which would be misleading.

`_log` does check, if a message is provided or not:

```bash
  if [[ -z ${2+set} ]]
  then
    echo "Call to '_log' is missing a message to log" >&2
    return 1
  fi
```

This error message is more clear than `parameter null or not set` which `${1:?}` produces if  $1 is empty or not set.

georglauterbach replied:
Hmm, true as well. I will double-check and take care!

georglauterbach replied:
Done in 1cbde1d139f519716d0c439eb871fe7c37ded85d. I added an explicit message for better debugging.

georglauterbach commented on target/scripts/helpers/log.sh:
```diff
@@ -1,5 +1,103 @@
+    echo -e "${MESSAGE}"
+  fi
+
+  return 0
```
Yes, but good style. I like ending the log with this statement so there is no way this function could / should fail. I hope this is not a blocker :)

casperklein replied:
A function returns the exit code of the last command. Consider the rare case, when "echo" fails to output the message. `return 0` would be misleading, as it indicates a successful function call, when the core purpose (output something) of the function failed.

IMO, this is not a good practice, to always end a function with `return 0`.

Not a blocker for me, but you might reconsider it, after reading my comment.

georglauterbach replied:
You're right - I was actually just thinking about the same. I will remove the line.

georglauterbach replied:
Done in [`1cbde1d`](https://github.com/docker-mailserver/docker-mailserver/pull/2493/commits/1cbde1d139f519716d0c439eb871fe7c37ded85d).

georglauterbach commented:
Sorry for the force-push, but I forgot one tiny change. Only "pushed" over one commit though. I applied the docs changes @polarathene as @casperklein also made me aware of them :) Some minor adjustments on the way (resolved many of the suggestions along the way).

There were some comments about the log format, but I took great care when crafting it so I would not like to change it unless there are very good reasons for it. @casperklein I left three suggestions open because I did not want to resolve them on my own. I did not change anything (as it was not necessary), but I wanted you to read them and resolve them if you feel appropriate.

I hope all is well now with this PR :D

georglauterbach commented:
Test in [`2b9ab7c`](https://github.com/docker-mailserver/docker-mailserver/pull/2493/commits/2b9ab7c52949678ccdfc773a65744015c8cc989d) failed because I added `' '` around a name. I corrected this in [`1cbde1d`](https://github.com/docker-mailserver/docker-mailserver/pull/2493/commits/1cbde1d139f519716d0c439eb871fe7c37ded85d) and streamlines the message to the new format (without dots at the end, etc.).

Some small adjustments because of PR suggestions happened with the last three commits, nothing big. Tests should pass now. The `start-mailserver.sh` messages are now normal `_log 'info' ...` messages, as discussed in https://github.com/docker-mailserver/docker-mailserver/pull/2493#discussion_r830544229.
